src/metabase/models/field_usage.clj

```diff
@@ -0,0 +1,87 @@
+(ns metabase.models.field-usage
+  (:require
+   [metabase.lib.core :as lib]
+   [metabase.lib.schema :as lib.schema]
+   [metabase.lib.util :as lib.util]
+   [metabase.models.interface :as mi]
+   [metabase.query-processor.middleware.fetch-source-query :as fetch-source-query]
+   [metabase.util.malli :as mu]
+   [methodical.core :as methodical]
+   [toucan2.core :as t2]
+   [toucan2.tools.disallow :as t2.disallow]))
+
+(doto :model/FieldUsage
+  (derive :metabase/model)
+  (derive ::t2.disallow/update)
+  (derive :hook/created-at-timestamped?))
+
+(methodical/defmethod t2/table-name :model/FieldUsage [_model] :field_usage)
+
+(t2/deftransforms :model/FieldUsage
+  {:used_in                   mi/transform-keyword
+   :aggregation_function      mi/transform-keyword
+   :breakout_temporal_unit    mi/transform-keyword
+   :breakout_binning_strategy mi/transform-keyword
+   :filter_op                 mi/transform-keyword})
+
+(defn- filter->field-usage
+  [query stage filter-clause]
+  (let [filter-parts (lib/filter-parts query stage filter-clause)
+        field-id     (-> filter-parts :column :id)]
+    (when (int? field-id)
+      {:field_id    field-id
+       :used_in     :filter
+       :filter_op   (-> filter-parts :operator :short)})))
+
+(defn- aggregation->field-usage
+  [query stage aggregation-clause]
+  (let [aggregation-column (lib/aggregation-column query stage aggregation-clause)
+        field-id           (:id aggregation-column)]
+    (when (int? field-id)
+      {:field_id             field-id
+       :used_in              :aggregation
+       :aggregation_function (first aggregation-clause)})))
+
+(defn- breakout->field-usage
```

Comment:
Up to you whether you want to make this change, but we do a lot of this "different logic for each kind of clause" in `metabase.lib.*`. Probably it would be

```clojure
(defmulti field-usage-method lib.dispatch/dispatch-value)

(defmethod field-usage-method :lib/aggregation [query stage aggregation-clause]
  ...)

(defn field-usage [query stage clause]
  (field-usage-method query stage clause))
```
and the defmethod can work for stages, queries, etc.; they're all different dispatch values.

You can see examples in lots of places, but maybe `display-info-method` or `metabase.lib.convert` are good examples.

As I say, this is your code so it's up to your team. Just wanted to make sure you were aware of the prior art.

Reply:
My initial implementation used defmulti, but we decided to go with pure functions because it's less code and we don't need the ability parse other types.

see [this](https://github.com/metabase/metabase/pull/39671#discussion_r1547548825) discussion for context